### DISCUSSION TITLE
sqlite3 header wrapper

### DISCUSSION THREAD

**Discussion by LukyGuyLucky** [c3lang/c3c#1163] [Q&A] [answered]
I am learning to make a sqlite3 wrapper header for C3 but failed.I have no clue what the compiler complained about.Attached the wrapper file for reference.And also attached all testing file in zip format.

[c3sqlite_test.zip](https://github.com/c3lang/c3c/files/14472052/c3sqlite_test.zip)

Acutally the testing wrapper just contains several lines(sqlite3.c3i):

module sqlite3;

def FnSqlite3Callback = fn int(void* parm1, int* parm2, char** parm3, char** parm4);
extern fn int sqlite3_open(char* filename, sqlite3** db);
extern fn int sqlite3_close(sqlite3* db);
extern fn int sqlite3_column_int(sqlite3_stmt* stmt, int parm2);
extern fn int sqlite3_prepare_v2(sqlite3* db, char* parm2, int parm3, sqlite3_stmt** parm4, char** parm5);
extern fn int sqlite3_step(sqlite3_stmt* stmt);
extern fn int sqlite3_finalize(sqlite3_stmt* stmt);
extern fn int sqlite3_exec(sqlite3* parm1, char* parm2, FnSqlite3Callback parm3, void* parm4, char** parm5);
extern fn void sqlite3_free(void* parm);

Below is the compiling message:

E:\Learning\c3pack\sql\c3sqlite>c3c compile main.c3 --lib sqlite3 --libdir . -l sqlite3.lib
 2:
 3: def FnSqlite3Callback = fn int(void* parm1, int* parm2, char** parm3, char** parm4);
 4:
 5: extern fn int sqlite3_open(char* filename, sqlite3** db);
                                                      ^
(E:/Learning/c3pack/sql/c3sqlite/_temp_build/_c3l/sqlite3.c3l_78248cc7/sqlite3.c3i:5:51) Error: Expected ')'.

 4:
 5: extern fn int sqlite3_open(char* filename, sqlite3** db);
 6:
 7: extern fn int sqlite3_close(sqlite3* db);
                                       ^
(E:/Learning/c3pack/sql/c3sqlite/_temp_build/_c3l/sqlite3.c3l_78248cc7/sqlite3.c3i:7:36) Error: Expected ')'.

 6:
 7: extern fn int sqlite3_close(sqlite3* db);
 8:
 9: extern fn int sqlite3_column_int(sqlite3_stmt* stmt, int parm2);
                                                 ^
(E:/Learning/c3pack/sql/c3sqlite/_temp_build/_c3l/sqlite3.c3l_78248cc7/sqlite3.c3i:9:46) Error: Expected ')'.

 8:
 9: extern fn int sqlite3_column_int(sqlite3_stmt* stmt, int parm2);
10:
11: extern fn int sqlite3_prepare_v2(sqlite3* db, char* parm2, int parm3, sqlite3_stmt** parm4, char** parm5);
                                            ^
(E:/Learning/c3pack/sql/c3sqlite/_temp_build/_c3l/sqlite3.c3l_78248cc7/sqlite3.c3i:11:41) Error: Expected ')'.

10:
11: extern fn int sqlite3_prepare_v2(sqlite3* db, char* parm2, int parm3, sqlite3_stmt** parm4, char** parm5);
12:
13: extern fn int sqlite3_step(sqlite3_stmt* stmt);
                                           ^
(E:/Learning/c3pack/sql/c3sqlite/_temp_build/_c3l/sqlite3.c3l_78248cc7/sqlite3.c3i:13:40) Error: Expected ')'.

12:
13: extern fn int sqlite3_step(sqlite3_stmt* stmt);
14:
15: extern fn int sqlite3_finalize(sqlite3_stmt* stmt);
                                               ^
(E:/Learning/c3pack/sql/c3sqlite/_temp_build/_c3l/sqlite3.c3l_78248cc7/sqlite3.c3i:15:44) Error: Expected ')'.

14:
15: extern fn int sqlite3_finalize(sqlite3_stmt* stmt);
16:
17: extern fn int sqlite3_exec(sqlite3* parm1, char* parm2, FnSqlite3Callback parm3, void* parm4, char** parm5);
                                      ^
(E:/Learning/c3pack/sql/c3sqlite/_temp_build/_c3l/sqlite3.c3l_78248cc7/sqlite3.c3i:17:35) Error: Expected ')'.
E:\Learning\c3pack\sql\c3sqlite>

Any help would be appreciated.

Regards,
Peter

**Comment by lerno** [comment]
C3 types must start with an uppercase character, so Sqlite3_stmt works (or just Stmt, recall there is proper namespacing) but sqlite3_stmt doesn't.

**Reply by LukyGuyLucky** [reply]
Got it.But sorry again,what's wrong with sqlite3* ?After modified the sqlite3.c3i by replace sqlite3_stmt with Sqlite3_stmt,the compiler produced the same error message to sqlite3 now.When I tried to modify sqlite3 to Sqlite3,it doesn't work as the compiler complains failed to found it.

 5: extern fn int sqlite3_open(char* filename, sqlite3** db);
                                                                                 ^
(E:/Learning/c3pack/sql/c3sqlite/_temp_build/_c3l/sqlite3.c3l_78248cc7/sqlite3.c3i:5:51) Error: Expected ')'.

**Reply by lerno** [reply]
How did you define the Sqlite3 struct?

**Reply by LukyGuyLucky** [reply]
Now I tried 
struct Sqlite3;//complier says no
try then
struct Sqlite3{} //compiler says no found Sqlite3_stmt

try then 
struct Sqlite3{}
struct Sqlite3_stmt{}

This time the compiler complains zero sized structs are not permitted.
(E:/Learning/c3pack/sql/c3sqlite/_temp_build/_c3l/sqlite3.c3l_78248cc7/sqlite3.c3i:5:8) Error: Zero sized strut not permitted.


The sqlite3.c3i ---

module sqlite3;


struct Sqlite3{}
struct Sqlite3_stmt{}
def FnSqlite3Callback = fn int(void* parm1, int* parm2, char** parm3, char** parm4);
extern fn int sqlite3_open(char* filename, Sqlite3** db);
extern fn int sqlite3_close(Sqlite3* db);
extern fn int sqlite3_column_int(Sqlite3_stmt* stmt, int parm2);
extern fn int sqlite3_prepare_v2(Sqlite3* db, char* parm2, int parm3, Sqlite3_stmt** parm4, char** parm5);
extern fn int sqlite3_step(Sqlite3_stmt* stmt);
extern fn int sqlite3_finalize(Sqlite3_stmt* stmt);
extern fn int sqlite3_exec(Sqlite3* parm1, char* parm2, FnSqlite3Callback parm3, void* parm4, char** parm5);
extern fn void sqlite3_free(void* parm);

**Reply by lerno** [reply]
Yes, so for structs that have no definition and are only used by pointer I recommend something like `distinct Sqlite3Ref = void*;` to define it. And then use Sqlite3Ref wherever you see `sqlite*`. `struct sqlite3_stmt*` => `distinct Sqlite3StmtRef = void*` etc

**Reply by LukyGuyLucky** [reply]
Great!A very big progress to me!

My sqlite3 exercise in C3 works now.

Thank you so much!

**Comment by LukyGuyLucky** [comment]
I have one more question:
In sqlite3.c3i,add
const SQLITE_OK=0;
If I place this .c3i file in sqlite3.c3l,and call if(sqlite3::SQLITE_OK) from main.c3,the compiler complains no found SQLITE_OK ;
but if I place the sqlite3.c3i file outside the sqlite3.c3l,say,just forget c3l package and direct compile with:
c3c compile-run main.c3  -l sqlite3.lib sqlite3.c3i
The program will be compiled and runs as expected.

So my question is how to properly define global const in Module A and call from Module B if Module A is inside c3l package?

Thanks again.

Regards,
Peter

**Reply by lerno** [reply]
To use a library like sqlite3.c3l, you need the following:

1. It is placed in a c3 library search path, `--libdir <dir>` allows you to set this from the command line.
2. You explicitly include the library you want (they are not added automatically). Use `--lib name` for this.

In your case: `c3c compile-run main.c3 -lib sqlite3` might be sufficient, if sqlite3.c3l is in the current directory.

**Reply by LukyGuyLucky** [reply]
Thanks.
Yes,for point 1,I am getting familar,this time the case(my question) is the compiler complains failed to find sqlite3::SQLITE_OK const:

module sqlite3;
const SQLITE_OK=0;
const SQLITE_ERROR=1;

module main;
import sqlite3;
fn void main(){if(sqlite3::SQLITE_OK){}}//can't find sqlite3::SQLITE_OK if sqlite3.c3i is in sqlite3.c3i

But it compiles in your point 2.

**Reply by lerno** [reply]
If you use `--lib sqlite3` then you should be able to access the constant like `sqlite3::SQLITE_OK` like in your example code. You should never include sqlite3.c3i explicitly.

**Reply by LukyGuyLucky** [reply]
No,it doesn't work,that's why I am asking help.
E:\Learning\c3pack\sql\c3sqlite>c3c compile-run main.c3 --lib sqlite3 --libdir . -l sqlite3
23:     char* path="world.db3";
24:     Sqlite3Ref db;
25:     int ret=sqlite3::sqlite3_open(path,&db);
26:     if (ret!=sqlite3::SQLITE_OK)
                          ^^^^^^^^^
(E:/Learning/c3pack/sql/c3sqlite/main.c3:26:20) Error: 'sqlite3::SQLITE_OK' could not be found, did you spell it right?

 30:    printf("Open database successfully.\n");
 31:
 32:    ret=sqlite3::sqlite3_close(db);
 33:    if (ret!=sqlite3::SQLITE_OK)
                          ^^^^^^^^^
(E:/Learning/c3pack/sql/c3sqlite/main.c3:33:20) Error: 'sqlite3::SQLITE_OK' could not be found, did you spell it right?

**Reply by lerno** [reply]
Could you perhaps provide a copy of your project and I could have a look?

**Reply by LukyGuyLucky** [reply]
Sure.Thanks.
[c3sqlite.zip](https://github.com/c3lang/c3c/files/14479385/c3sqlite.zip)

**Reply by lerno** [reply, marked as answer]
Your package works for me:

```
/Users/lerno/Projects/c3c/cmake-build-debug-16/c3c compile-run main.c3 --lib sqlite3
Program linked to executable 'main'.
Launching main...
Attempting to open database...
Open database successfully.
db closed.
```

**Reply by LukyGuyLucky** [reply]
I restrutured the whole folder (there are several other testing files inside the project folder) and try to compile,yes,it compiles.
Sorry for my careless.

Thank you so much!

Regards,
Peter

**Reply by lerno** [reply]
I'm happy it worked out!
